src/js/view/floating/floating-controller.ts

```diff
@@ -0,0 +1,327 @@
+            const fm = this.getFloatMode();
+            if (fm === 'notVisible') {
+                if (this._isMobile) {
+                    viewsManager.addScrollHandler(this._boundThrottledMobileFloatScrollHandler);
+                    this._boundThrottledMobileFloatScrollHandler();
+                } else {
+                    this.checkFloatIntersection();
+                }
+            } else if (fm === 'always') {
+                this.startFloating();
+            } else if (fm === 'never') {
+                this.stopFloating();
+            }
+        }
+    }
+    updatePlayerBounds(pb: BoundsObj): void {
+        this._playerBounds = pb;
+    }
+    getFloatingConfig(): FloatConfig | undefined {
+        return this._model.get('floating');
+    }
+    getFloatMode(): string {
+        const fc = this.getFloatingConfig();
+        return (fc && fc.mode) || 'notVisible';
+    }
+    resize(): void {
+        if (this._model.get('isFloating')) {
+            this.updateFloatingSize();
+        }
+    }
+    fosMobileBehavior(): boolean {
+        return this._isMobile && !deviceIsLandscape() && !this._model.get('fullscreen');
+    }
+    shouldFloatOnViewable(): boolean {
+        const state = this._model.get('state');
+        return state !== STATE_IDLE && state !== STATE_ERROR && state !== STATE_COMPLETE;
+    }
+    startFloating(mobileFloatIntoPlace?: boolean): void {
+        const playerBounds = this._playerBounds;
+        if (_floatingPlayer === null) {
+            _floatingPlayer = this._playerEl;
+
+            this._model.set('isFloating', true);
+
+            addClass(this._playerEl, 'jw-flag-floating');
+
+            if (mobileFloatIntoPlace) {
+                // Creates a dynamic animation where the top of the current player
+                // Smoothly transitions into the expected floating space in the event
+                // we can't start floating at 62px
+                style(this._wrapperEl, {
+                    transform: `translateY(-${FLOATING_TOP_OFFSET - playerBounds.top}px)`
+                });
+
+                setTimeout(() => {
+                    style(this._wrapperEl, {
+                        transform: 'translateY(0)',
+                        transition: 'transform 150ms cubic-bezier(0, 0.25, 0.25, 1)'
+                    });
+                });
+            }
+
+            // Copy background from preview element, fallback to image config.
+            const previewEl = this._preview.el as HTMLElement;
+            style(this._playerEl, {
+                backgroundImage: previewEl.style.backgroundImage || this._model.get('image')
+            });
+
+            this.updateFloatingSize();
+
+            if (!this._model.get('instreamMode')) {
+                this._floatingUI.enable();
+            }
+
+            // Perform resize and trigger "float" event responsively to prevent layout thrashing
+            this._model.trigger('forceResponsiveListener', {});
```

Comment:
interesting. I guess it's the view model so sure! better than having even more event dispatchers.